datafusion/functions/src/datetime/to_time.rs

```diff
@@ -0,0 +1,582 @@
+    ) -> Result<ColumnarValue> {
+        let args = args.args;
+        if args.is_empty() {
+            return exec_err!("to_time function requires 1 or more arguments, got 0");
+        }
+
+        // validate that any args after the first one are Utf8
+        if args.len() > 1 {
+            validate_data_types(&args, "to_time")?;
+        }
+
+        match args[0].data_type() {
+            Utf8View | LargeUtf8 | Utf8 => self.to_time(&args),
+            Null => Ok(ColumnarValue::Scalar(ScalarValue::Time64Nanosecond(None))),
+            // Support timestamp input by extracting time portion
+            Timestamp(_, _) => {
+                let nanos = extract_time_from_timestamp(&args[0])?;
+                Ok(nanos)
+            }
+            other => {
+                exec_err!("Unsupported data type {} for function to_time", other)
+            }
+        }
+    }
+
+    fn documentation(&self) -> Option<&Documentation> {
+        self.doc()
+    }
+}
+
+/// Parse time array using the provided formats
+fn parse_time_array<'a, A: StringArrayType<'a>>(
+    array: &A,
+    formats: &[&str],
+) -> Result<PrimitiveArray<Time64NanosecondType>> {
+    let mut builder: PrimitiveBuilder<Time64NanosecondType> =
+        PrimitiveArray::builder(array.len());
+
+    for i in 0..array.len() {
+        if array.is_null(i) {
+            builder.append_null();
+        } else {
+            let s = array.value(i);
+            let nanos = parse_time_with_formats(s, formats)?;
+            builder.append_value(nanos);
+        }
+    }
+
+    Ok(builder.finish())
+}
+
+/// Parse time string using provided formats
+fn parse_time_with_formats(s: &str, formats: &[&str]) -> Result<i64> {
+    for format in formats {
+        if let Ok(time) = NaiveTime::parse_from_str(s, format) {
+            return Ok(time_to_nanos(time));
+        }
+    }
+    exec_err!(
+        "Error parsing '{}' as time. Tried formats: {:?}",
+        s,
+        formats
+    )
+}
+
+/// Convert NaiveTime to nanoseconds since midnight
+fn time_to_nanos(time: NaiveTime) -> i64 {
+    let hours = time.hour() as i64;
+    let minutes = time.minute() as i64;
+    let seconds = time.second() as i64;
+    let nanos = time.nanosecond() as i64;
+
+    hours * NANOS_PER_HOUR
+        + minutes * NANOS_PER_MINUTE
+        + seconds * NANOS_PER_SECOND
+        + nanos
+}
+
+/// Extract time portion from timestamp (nanoseconds since midnight)
+fn extract_time_from_timestamp(arg: &ColumnarValue) -> Result<ColumnarValue> {
+    match arg {
+        ColumnarValue::Scalar(scalar) => {
+            let nanos = match scalar {
+                ScalarValue::TimestampNanosecond(Some(ts), _) => *ts % NANOS_PER_DAY,
+                ScalarValue::TimestampMicrosecond(Some(ts), _) => {
+                    (*ts * 1_000) % NANOS_PER_DAY
+                }
+                ScalarValue::TimestampMillisecond(Some(ts), _) => {
+                    (*ts * 1_000_000) % NANOS_PER_DAY
+                }
```

Comment:
Same here; in this case we can create arrays of size 1 to pass into the arrow cast kernel and recreate a scalar value from the output, to reuse it's code